Sources/NIOAsyncAwaitServerExample/Server.swift

```diff
@@ -0,0 +1,175 @@
+        //
+        // Note that this does not spawn a background Task in order to do this:
+        // it takes over whatever Task it is spawned on. This is an important
+        // property: it allows the user to decide which Task tree is responsible
+        // for running the server, and therefore responsible for cancelling it.
+        //
+        // In here we create a task group of our own: the way we intend to use it
+        // is discussed below.
+
+        // Note that we start this group with 1 thread. In general, most NIO programs
+        // should use 1 thread as a default unless they're planning to be a network
+        // proxy or something that is expecting to be dominated by packet parsing. Most
+        // servers aren't, and NIO is very fast, so 1 NIO thread is quite capable of
+        // saturating the average small to medium sized machine.
+        let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+
+        let serverChannelAndStream: ChannelAndStream<Channel>
+        do {
+            // Inside this do block we want to catch any errors from creating the bootstrap.
+            // We still have to tidy up, so we can't just let this throw from here.
+            serverChannelAndStream = try await self.makeServerChannel(group: group)
+        } catch {
+            // Couldn't create the server channel. We're done here. Shut down the loop and go home.
+            try await group.shutdownGracefully()
+            throw error
+        }
+
+        // We split these two apart here to avoid capturing more than we want to
+        // in any given closure.
+        let serverChannel = serverChannelAndStream.channel
+        let serverChannelStream = serverChannelAndStream.stream
+
+        // We now have the server channel and the stream, we can start our work. First, though,
+        // we need to prepare to handle cancellation.
+        //
+        // If this Task is cancelled, the first thing to do is to shut down the server channel.
+        // This will have the effect of terminating the async loop that makes up the rest of this Task,
+        // which will naturally bring it to a close.
+
+        // TODO: error handling!
+        try await withTaskCancellationHandler {
+            // Now we accept new connections in a loop. Each of these is going to be dispatched
+            // to a new child Task in a TaskGroup. These Tasks will be responsible for running the
+            // child connections.
+            try await withThrowingTaskGroup(of: Void.self) { taskGroup in
+                for try await childChannel in serverChannelStream {
+                    taskGroup.addTask {
+                        await Self.handleChildChannel(childChannel)
+                    }
+                }
+            }
+        } onCancel: {
+            // This is a nice property of NIO: we can do this without actually
+            // awaiting it.
+            serverChannel.close(promise: nil)
+        }
```

Comment:
What do you mean by "async cancellation"

Reply:
Arguably the reason people want to await in `onCancel` is because cancel work in an actor; and the ACTUAL missing feature here is `send`, i.e. "notify that actor that <something> but no need to await" that's the semantics we need for such cancellation handlers IMHO, and we'll keep hitting that wall until we get "don't await" (i.e. `send`)

Reply:
Ah, that makes sense. Thanks for elaborating! What about errors that can happen during cancellation, should they just be dropped?

Reply:
You'd need some "back channel"; generally there's too much complexity in this and in reality... what would you even do when cancelling something failed? Cancelling is always racy and best effort to begin with, so it's not a reliable mechanism you can build upon for "did it definitely cancel or not" etc. 

This is similar to why we decided that reactive streams have no channel for cancellation -- all you can to is send completion/error _forward_ but if the cancel even had an effect or not, you're never 100% sure unless you exactly know what error it should cause. I.e. the cancellation after all propagates to children in our system, so the child task may return/complete and you don't really know as the outsider (NIO) if that just completed quickly, or BECAUSE of the cancellation that some child task checked for.

Reply:
This is something that I've ended up wrestling with. Often getting stuck with not being able to call actor (async) functions from within `deinit` or the Task `cancellationHandler`. I think @ktoso comments are sensible, there's only so much you can do, short of spinning off an unstructured `Task` within one of these contexts, which I think undermines the idea of structured concurrency that we're aiming for here?

Reply:
Yeah, what Cory said -- we're missing language features IMHO, there should be a `send actor.cancelPlease()` that _does not_ await, and could have a slightly lighter task, but that also would be spawning a new task to kick off the work.